sklearn/manifold/tests/test_t_sne.py

```diff
@@ -806,3 +813,22 @@ def test_tsne_with_different_distance_metrics():
             metric='precomputed', n_components=n_components_embedding,
             random_state=0).fit_transform(dist_func(X))
         assert_array_equal(X_transformed_tsne, X_transformed_tsne_precomputed)
+
+
+def test_tsne_with_different_csr_distance_metrics():
+    """Make sure that TSNE works for different csr distance metrics"""
+    random_state = check_random_state(0)
+    n_components_original = 3
+    n_components_embedding = 2
+    X = random_state.randn(50, n_components_original).astype(np.float32)
+    X_csr = sp.csr_matrix(X)
+    metrics = ['manhattan', 'cosine']
+    dist_funcs = [manhattan_distances, cosine_distances]
+    for metric, dist_func in zip(metrics, dist_funcs):
+        X_transformed_tsne = TSNE(
+            metric=metric, n_components=n_components_embedding,
+            random_state=0).fit_transform(X_csr)
+        X_transformed_tsne_precomputed = TSNE(
+            metric='precomputed', n_components=n_components_embedding,
+            random_state=0).fit_transform(dist_func(X_csr))
```

Comment:
This does not produce a sparse distance matrix. You should produce a sparse distance matrix with `kneighbors_graph`.

Reply:
I am not clear with this. An example would suffice.

Reply:
Can you give an example in this case? I am not very clear in how to achieve this

Reply:
For arbitrary X:
```python
D = pairwise_distances(X)
D5 = neighbors.kneighbors_graph(X, 5)
```
D5 is sparse precomputed version of D that should work for `n_neighbors <= 5`. You can also split it into train and test with `D5[:n_train, :n_train]`, `D5[n_train:, :n_train]`.

Reply:
This raises a mismatch error. Can you please tell where I am going wrong. (Also tried changing the parameter of `kneighbors.graph` to `X_csr`, still a mismatch)
```Python
for metric, dist_func in zip(metrics, dist_funcs):
        X_transformed_tsne = TSNE(
            metric=metric, n_components=n_components_embedding, perplexity=5,
            random_state=0).fit_transform(X_csr)
        temp = dist_func(X_csr)
        temp2 = neighbors.kneighbors_graph(temp, 40)
        X_transformed_tsne_precomputed = TSNE(
            metric='precomputed', n_components=n_components_embedding, perplexity=5,
            random_state=0).fit_transform(temp2)
        assert_array_equal(X_transformed_tsne, X_transformed_tsne_precomputed)
```

Reply:
`kneighbors_graph` is called on `X`, not `D`. But I made a mistake: you need to call `kneighbors_graph(mode='distance')` (I wish this were the default). Also, you should probably be comparing that sparse precomputed gets the same results as dense precomputed, not comparing sparse precomputed to sparse raw here the raw-precomputed invariance and the non-precomputed saprse-dense invariance should be tested separately.

But the test will still fail, because there is a bug in your changes to NearestNeighbors. The precomputed sparse and dense input do not produce the same results.

When I use dense precomputed (for `X, _ = make_blobs(random_state=0)`), the kneighbors call in tsne produces these indices:
```
[[ 2 41 56  8]
 [52 80  5 50]
 [56  0  8 41]
 [29 92 83 52]
 [17 95  7 23]
 [50 36 86 81]
 [94 62 19 14]
 [70 95 49  4]
 [38 88 25  0]
...
```

```
[[ 41.  56.   8.  25.]
 [ 80.   5.  50.  83.]
 [  0.   8.  41.  25.]
 [ 92.  83.  52.  70.]
 [ 95.   7.  23.  87.]
 [ 36.  86.  81.  89.]
 [ 62.  19.  14.  61.]
 [ 95.  49.   4.  92.]
 [ 88.  25.   0.   2.]
 [ 45.  89.  50.  52.]
```
Firstly, you can see the dtype is incorrect on sparse input. Secondly you can see that the last three columns of the first results match the first three columns of the second results, so there seems to be some kind of off-by-one logic error.

Reply:
`kneighbors_graph` is called on `X`, not `D`. But I made a mistake: you need to call `kneighbors_graph(mode='distance')` (I wish this were the default). Also, you should probably be comparing that sparse precomputed gets the same results as dense precomputed, not comparing sparse precomputed to sparse raw here the raw-precomputed invariance and the non-precomputed saprse-dense invariance should be tested separately.

But the test will still fail, because there is a bug in your changes to NearestNeighbors. The precomputed sparse and dense input do not produce the same results.

When I use dense precomputed (for `X, _ = make_blobs(random_state=0)`), the kneighbors call in tsne produces these indices:
```
[[ 2 41 56  8]
 [52 80  5 50]
 [56  0  8 41]
 [29 92 83 52]
 [17 95  7 23]
 [50 36 86 81]
 [94 62 19 14]
 [70 95 49  4]
 [38 88 25  0]
...
```

```
[[ 41.  56.   8.  25.]
 [ 80.   5.  50.  83.]
 [  0.   8.  41.  25.]
 [ 92.  83.  52.  70.]
 [ 95.   7.  23.  87.]
 [ 36.  86.  81.  89.]
 [ 62.  19.  14.  61.]
 [ 95.  49.   4.  92.]
 [ 88.  25.   0.   2.]
 [ 45.  89.  50.  52.]
```
Firstly, you can see the dtype is incorrect on sparse input. Secondly you can see that the last three columns of the first results match the first three columns of the second results, so there seems to be some kind of off-by-one logic error.

Reply:
The error in dtype is resolved. But locally I am getting the same neighbors indices in TSNE.
I ran the following snippet after adding a print statement after kneighbors call.
```Python
X, _ = make_blobs(n_samples=16, n_features=16, random_state = 0)
X = abs(X)
tsne = TSNE(metric="precomputed")
tsne.fit_transform(X)

X1 = csr_matrix(X)
tsne.fit_transform(X1)
```
```Python
[[ 1  3  5 12 10  4  9 13  8  7 14 15 11  2  6]
 [ 9  0  8 13  4  5 10  3  7 12 14 15  6 11  2]
 [ 0 12 11  4  5  6  9  3  1 10 13  8 15 14  7]
 [12  6  9 13 15  5 14  7  2  1 10  0  8 11  4]
 [ 0  9  6 11  2  3 12  5  1 10 13 15  7  8 14]
 [12 11  0  6  9  4  3  2  1 10 15 13 14  7  8]
 [12 13  9 14  8  2  7  3  5  1 15 10  4  0 11]
 [ 3  0  4  9  6 11  2 12  5 10  1 15 14 13  8]
 [ 6 12 13  9  5  7 15 14 10  1  2  3 11  0  4]
 [ 8  7  1  4  5  3  0 13 10 12 15 14  2 11  6]
 [13 12  9  6 14  2  1  7 15  3  5  8 11  4  0]
 [12  3  0  2  6  4  5  9 10  1  7 14 15 13  8]
 [ 5  0  9 13  1  3  8  4 10  7 14 15 11  6  2]
 [12  6 15  9  5  7  2 14  1 10  3  8 11  0  4]
 [ 3 11  9  6  0  5  4  2 12  1 10  7  8 15 13]
 [ 0  9  4  1  5  3 12 13 10  7  8 14  2  6 11]]
```
```Python
[[ 1  3  5 12 10  4  9 13  8  7 14 15 11  2  6]
 [ 9  0  8 13  4  5 10  3  7 12 14 15  6 11  2]
 [ 0 12 11  4  5  6  9  3  1 10 13  8 15 14  7]
 [12  6  9 13 15  5 14  7  2  1 10  0  8 11  4]
 [ 0  9  6 11  2  3 12  5  1 10 13 15  7  8 14]
 [12 11  0  6  9  4  3  2  1 10 15 13 14  7  8]
 [12 13  9 14  8  2  7  3  5  1 15 10  4  0 11]
 [ 3  0  4  9  6 11  2 12  5 10  1 15 14 13  8]
 [ 6 12 13  9  5  7 15 14 10  1  2  3 11  0  4]
 [ 8  7  1  4  5  3  0 13 10 12 15 14  2 11  6]
 [13 12  9  6 14  2  1  7 15  3  5  8 11  4  0]
 [12  3  0  2  6  4  5  9 10  1  7 14 15 13  8]
 [ 5  0  9 13  1  3  8  4 10  7 14 15 11  6  2]
 [12  6 15  9  5  7  2 14  1 10  3  8 11  0  4]
 [ 3 11  9  6  0  5  4  2 12  1 10  7  8 15 13]
 [ 0  9  4  1  5  3 12 13 10  7  8 14  2  6 11]]

```

Could you please point out where I am going wrong here?

Reply:
With the latest version, I find this passes:
```python
import numpy as np
from sklearn.datasets import make_blobs
from scipy.sparse import csr_matrix
from sklearn.metrics import pairwise_distances
from sklearn.manifold import TSNE
X, _ = make_blobs(n_samples=16, n_features=16, random_state=0)
X = abs(X)
tsne = TSNE(metric="precomputed", random_state=0)
Xt = tsne.fit_transform(X)

X1 = csr_matrix(X)
Xt1 = tsne.fit_transform(X1)
assert np.allclose(Xt, Xt1)
```
Do you find it does not pass?

Reply:
This does pass.
But still we are not done with modifying the test function. This is a complete working snippet of the problem. Can you figure out why this is not working still? I don't find off by one logic error :(

```Python
import numpy as np
from scipy.sparse import csr_matrix
from sklearn.utils import check_random_state
from sklearn.manifold.t_sne import TSNE
from sklearn.metrics.pairwise import manhattan_distances
from sklearn.metrics.pairwise import cosine_distances
from sklearn.neighbors import kneighbors_graph
from sklearn.utils.testing import assert_array_equal

def test_tsne_with_different_csr_distance_metrics():
    """Make sure that TSNE works for different csr distance metrics"""
    random_state = check_random_state(0)
    n_components_original = 3
    n_components_embedding = 2
    X = random_state.randn(50, n_components_original).astype(np.float32)
    X_csr = csr_matrix(X)
    metrics = ['manhattan', 'cosine']
    dist_funcs = [manhattan_distances, cosine_distances]
    for metric, dist_func in zip(metrics, dist_funcs):
        X_transformed_tsne = TSNE(
            metric=metric, n_components=n_components_embedding, perplexity=5,
            random_state=0).fit_transform(X)
        D = dist_func(X)
        D_csr = kneighbors_graph(X, 19, mode='distance')
        X_transformed_tsne_precomputed = TSNE(
            metric='precomputed', n_components=n_components_embedding, perplexity=5,
            random_state=0).fit_transform(D_csr)
        print(X_transformed_tsne, X_transformed_tsne_precomputed)
        assert_array_equal(X_transformed_tsne, X_transformed_tsne_precomputed)

test_tsne_with_different_csr_distance_metrics()
```
Sorry for not providing complete code before.

Reply:
The bug is that you have not taken into account the `query_is_train` logic in kneighbors when handling sparse precomputed input. And it may not be trivial to do so.

Reply:
Can you please provide a script to reproduce error mentioned here:
https://github.com/scikit-learn/scikit-learn/pull/10206#discussion_r157468605

Reply:
```python
import numpy as np
from sklearn.datasets import make_blobs
from scipy.sparse import csr_matrix
from sklearn.metrics import pairwise_distances
from sklearn.manifold import TSNE
from sklearn.neighbors import NearestNeighbors, kneighbors_graph


def test():
    X, _ = make_blobs(n_samples=16, n_features=16, random_state=0)
    nn = NearestNeighbors(metric='precomputed', n_neighbors=5)
    nn.fit(pairwise_distances(X))
    dist1, neigh1 = nn.kneighbors()

    nn.fit(kneighbors_graph(X, n_neighbors=15, mode='distance'))
    dist2, neigh2 = nn.kneighbors()
    assert np.all(neigh1 == neigh2)
    assert np.allclose(dist1, dist2)
```

run with `pytest -vv /path/to/test.py`

Reply:
Added the logic with `query_is_train`. I guess now I also need to add a warning in case of invalid distance matrix because it raises error if distance matrix is invalid.

Still the above https://github.com/scikit-learn/scikit-learn/pull/10206#discussion_r157671201 function here raises a mismatch. What needs to be modified now?